src/declarativedbusinterface.cpp

```diff
@@ -173,102 +175,268 @@ void DeclarativeDBusInterface::call(const QString &method, const QJSValue &argum
+        arr << qvariant_cast<T>(var);
+    }
+    return arr;
+}
+
+static QStringList toQStringList(const QVariantList &lst) {
+    QStringList arr;
+    foreach(const QVariant &var, lst) {
+        arr << qvariant_cast<QString>(var);
+    }
+    return arr;
+};
+
+static QByteArray toQByteArray(const QVariantList &lst) {
+    QByteArray arr;
+    foreach(const QVariant &var, lst) {
+        uchar tmp = static_cast<uchar>(var.toUInt());
+        arr.append(static_cast<char>(tmp));
+    }
+    return arr;
+};
+
+static void registerDBusTypes(void)
+{
+    static bool done = false;
+
+    if( !done ) {
+        done = true;
+
+        qDBusRegisterMetaType< QList<bool> >();
+        qDBusRegisterMetaType< QList<int> >();
+        qDBusRegisterMetaType< QList<double> >();
+
+        qDBusRegisterMetaType< QList<quint8> >();
+        qDBusRegisterMetaType< QList<quint16> >();
+        qDBusRegisterMetaType< QList<quint32> >();
+        qDBusRegisterMetaType< QList<quint64> >();
+
+        qDBusRegisterMetaType< QList<qint16> >();
+        qDBusRegisterMetaType< QList<qint32> >();
+        qDBusRegisterMetaType< QList<qint64> >();
+    }
+}
+
+static bool flattenVariantList(QVariant &var, const QVariantList &lst,
+                               int typeChar)
+{
+    bool res = true;
+
+    switch( typeChar ) {
+    case 'b': // BOOLEAN
+        var = QVariant::fromValue(toQList<bool>(lst));
+        break;
+    case 'y': // BYTE
+        var = QVariant::fromValue(toQByteArray(lst));
+        break;
+    case 'q': // UINT16
+        var = QVariant::fromValue(toQList<quint16>(lst));
+        break;
+    case 'u': // UINT32
+        var = QVariant::fromValue(toQList<quint32>(lst));
+        break;
+    case 't': // UINT64
+        var = QVariant::fromValue(toQList<quint64>(lst));
+        break;
+    case 'n': // INT16
+        var = QVariant::fromValue(toQList<qint16>(lst));
+        break;
+    case 'i': // INT32
+        var = QVariant::fromValue(toQList<qint32>(lst));
+        break;
+    case 'x': // INT64
+        var = QVariant::fromValue(toQList<qint64>(lst));
+        break;
+    case 'd': // DOUBLE
+        var = QVariant::fromValue(toQList<double>(lst));
+        break;
+    case 's': // STRING
+        var = QVariant::fromValue(toQStringList(lst));
+        break;
+    default:
+        res = false;
+        break;
+    }
+
+    return res;
+}
 
-template<typename T>
-T extractTypedList(const QVariantList &list)
+static bool flattenVariantArrayForceType(QVariant &var, int typeChar)
 {
-    T rv;
-    foreach (const QVariant &element, list) {
-        rv.append(element.value<typename T::value_type>());
+    return flattenVariantList(var, var.toList(), typeChar);
+}
+
+static void flattenVariantArrayGuessType(QVariant &var)
+{
+    /* If the value can't be converted to a variant list
+     * or if the resulting list would be empty: use the
+     * value without modification */
+    QVariantList arr = var.toList();
+    if( arr.empty() )
```

Comment:
Nitpicking: part of the code use spaces between an expression and spaces, whereas some parts do not. E.g.

> if ( expr ) ;

vs

> if (x) ;